android/app/src/main/java/org/fossasia/openevent/adapters/viewholders/SessionViewHolder.java

```diff
@@ -1,66 +1,307 @@
+        }
+
+    }
+
+    private void handleVisibilityByType(int type) {
+        switch (type) {
+            case trackWiseSessionList:
+                trackImageIcon.setVisibility(View.GONE);
+                sessionTrack.setVisibility(View.GONE);
+                break;
+
+            case locationWiseSessionList:
+                sessionLocation.setVisibility(View.GONE);
+                locationIcon.setVisibility(View.GONE);
+                break;
+
+            case speakerWiseSessionList:
+                sessionSpeaker.setVisibility(View.GONE);
+                speakerIcon.setVisibility(View.GONE);
+                break;
+
+            default: // Shouldn't reach here
+        }
+    }
+
+    private void setBookmarkClickListener(RealmDataRepository realmRepo, Track track) {
+
+        final int sessionId = session.getId();
+        sessionBookmarkIcon.setOnClickListener(v -> {
+            if (track == null) return;
+
+            if (session.getIsBookmarked()) {
+
+                realmRepo.setBookmark(sessionId, false).subscribe();
+                setBookmarkIcon(
+                        sessionBookmarkIcon, false, track.getFontColor());
+
+                if (onBookmarkSelectedListener != null)
+                    onBookmarkSelectedListener.showSnackbar(new BookmarkStatus(Color.parseColor(track.getColor()),
+                            sessionId, BookmarkStatus.Status.CODE_UNDO_REMOVED));
+
+            } else {
+                NotificationUtil.createNotification(session, context).subscribe(
+                        () -> {
+                            if (onBookmarkSelectedListener != null)
+                                onBookmarkSelectedListener.showSnackbar(new BookmarkStatus(Color.parseColor(track.getColor()),
+                                        sessionId, CODE_UNDO_ADDED));
+                        },
+                        throwable -> {
+                            if (onBookmarkSelectedListener != null)
+                                onBookmarkSelectedListener.showSnackbar(new BookmarkStatus(-1, -1, CODE_ERROR));
+                        });
+
+                realmRepo.setBookmark(sessionId, true).subscribe();
+                setBookmarkIcon(
+                        sessionBookmarkIcon, true, track.getFontColor());
+            }
+            WidgetUpdater.updateWidget(context);
+        });
+    }
+
+    private void setBookmarkIcon(ImageView sessionBookmarkIcon, boolean bookmarked, String color) {
+        if (bookmarked) {
+            sessionBookmarkIcon.setImageResource(R.drawable.ic_bookmark_white_24dp);
+        } else {
+            sessionBookmarkIcon.setImageResource(R.drawable.ic_bookmark_border_white_24dp);
+        }
+        if (!Utils.isEmpty(color))
+            sessionBookmarkIcon.setColorFilter(Color.parseColor(color), PorterDuff.Mode.SRC_ATOP);
+    }
+
+    public void setSession(Session session) {
+        this.session = session;
```

Comment:
All the above functions should be in ViewModel

Reply:
setSession()? But this session is used in this ViewHolder only. I will have to pass the view model in the recylerview adapter then right?

Reply:
I meant all above logic

Reply:
@iamareebjamal I made an interface in `SessionListAdapter` which returns this :
    public interface LinkViewModel {
        Observable<String> getSpeakerList();
    }

Since the adapter can't have a view model, should I put the code in all the view models in which the adapter is used or make a common helper method?

Reply:
Why should this interface be needed? I think we are handling click in the adapter right, just create a listener in adapter which tells fragment which item was clicked and this information will be passed to viewmodel and it will handle the click

Reply:
Currently we are handling clicks in the view holder itself, ie, we are setting the `onClickListener` while setting up the UI and directly handling clicks from there like launching new activity etc. Clicks I will handle in the fragment only like we normally do.

I made that interface to handle this :
![image](https://user-images.githubusercontent.com/22395998/35700115-1135584e-07b8-11e8-9b18-4fb18162e7ee.png)

I am stuck with this method.